src/common/transformations/src/transformations/common_optimizations/shared_ops_optimization.cpp

```diff
@@ -51,6 +51,10 @@ bool shared_node_optimization(const shared_ptr<Model>& model, const rules_t& rul
                     continue;
                 const auto& are_equal = rules.at(item.first);
 
+                std::sort(shared_nodes.begin(), shared_nodes.end(), [](Node* l_node, Node* r_node) {
+                    return l_node->get_instance_id() < r_node->get_instance_id();
```

Comment:
I looks like the actual root cause is in `get_target_inputs` function
![image](https://github.com/openvinotoolkit/openvino/assets/21181928/fe7fe1f1-c233-4f16-86da-d1d9f1725ea4)
![image](https://github.com/openvinotoolkit/openvino/assets/21181928/74f98fc2-c183-4bff-9cfd-ec6c93b6082e)
It sorts the resulting Inputs by the address of the nodes
So maybe it's better to fix this function.
@jane-intel @evkotov what do you think?

Reply:
@ilyachur could you also comment this problem?

Reply:
I am afraid of changing basic functionality in library, since it could cause some side effects.

Reply:
Created [PR](https://github.com/openvinotoolkit/openvino/pull/20365) to check side effects of changing Input/Output comparison